### PR TITLE
phpPackages.composer: use `fetchgit`

### PR DESCRIPTION
**Pull request by drupol**
Using `fetchgit` prevent the `composer.lock` from being stripped out when `fetchFromGitHub` is being used

https://nixpk.gs/pr-tracker.html?pr=258281

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by drupol** [comment]
@GrahamcOfBorg build phpPackages.composer

**Comment by NetaliDev** [comment]
Result of `nixpkgs-review pr 258281` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>php82Packages.php-parallel-lint</li>
    <li>php83Packages.php-parallel-lint</li>
  </ul>
</details>
<details>
  <summary>23 packages built:</summary>
  <ul>
    <li>adminer</li>
    <li>bookstack</li>
    <li>pdepend</li>
    <li>php81Packages.box</li>
    <li>php81Packages.castor</li>
    <li>php81Packages.composer</li>
    <li>php81Packages.php-parallel-lint</li>
    <li>php81Packages.psysh</li>
    <li>php82Packages.box</li>
    <li>php82Packages.castor</li>
    <li>php82Packages.composer</li>
    <li>php82Packages.psysh</li>
    <li>php83Packages.box</li>
    <li>php83Packages.castor</li>
    <li>php83Packages.composer</li>
    <li>php83Packages.psysh</li>
    <li>phpactor</li>
    <li>phpdocumentor</li>
    <li>pixelfed</li>
    <li>platformsh</li>
    <li>robo</li>
    <li>snipe-it</li>
    <li>vimPlugins.phpactor</li>
  </ul>
</details>

Two packages fail to build, but that does not look like a composer problem:
```console
error: builder for '/nix/store/24837mgbrxkdqzx79m40pc8ya5fd1pp9-php-php-parallel-lint-1.3.2.drv' failed with exit code 1;
       last 10 log lines:
       >   Error Output:
       >   ================
       >   Failed loading /nix/store/kz8ahpqyvz7cyj52kz417g65bkz3rqjh-php-opcache-8.1.
       >   24/lib/php/extensions/opcache.so:  /nix/store/kz8ahpqyvz7cyj52kz417g65bkz3r
       >   qjh-php-opcache-8.1.24/lib/php/extensions/opcache.so: undefined symbol: zen
       >   d_build_delayed_early_binding_list
       >
       >
       > compile [--debug] [--no-parallel] [--no-restart] [--dev] [--no-config] [--with-docker] [--composer-bin COMPOSER-BIN] [--allow-composer-check-failure] [-c|--config CONFIG] [-d|--working-dir WORKING-DIR]
       >
       For full logs, run 'nix log /nix/store/24837mgbrxkdqzx79m40pc8ya5fd1pp9-php-php-parallel-lint-1.3.2.drv'.
error: builder for '/nix/store/66l7q0j6zvv8n4jky49nhzvbacczp80b-php-php-parallel-lint-1.3.2.drv' failed with exit code 1;
       last 10 log lines:
       >   Error Output:
       >   ================
       >   Failed loading /nix/store/kz8ahpqyvz7cyj52kz417g65bkz3rqjh-php-opcache-8.1.
       >   24/lib/php/extensions/opcache.so:  /nix/store/kz8ahpqyvz7cyj52kz417g65bkz3r
       >   qjh-php-opcache-8.1.24/lib/php/extensions/opcache.so: undefined symbol: zen
       >   d_build_delayed_early_binding_list
       >
       >
       > compile [--debug] [--no-parallel] [--no-restart] [--dev] [--no-config] [--with-docker] [--composer-bin COMPOSER-BIN] [--allow-composer-check-failure] [-c|--config CONFIG] [-d|--working-dir WORKING-DIR]
       >
       For full logs, run 'nix log /nix/store/66l7q0j6zvv8n4jky49nhzvbacczp80b-php-php-parallel-lint-1.3.2.drv'.
error: 2 dependencies of derivation '/nix/store/vm72zp3djc13zzynjh7bb955psnl4p2w-review-shell.drv' failed to build
```

**Comment by etu** [comment]
May be worth to leave a comment in the code as well if why fetchgit is used
rather than fethFromGitHub

**Comment by drupol** [comment]
Done.

**Comment by drupol** [comment]
@GrahamcOfBorg build phpPackages.composer

File: pkgs/development/php-packages/composer/default.nix [hunk 1]
```diff
@@ -1,16 +1,17 @@
...
 php.buildComposerProject (finalAttrs: {
   composer = callPackage ../../../build-support/php/pkgs/composer-phar.nix { };
 
   pname = "composer";
   version = "2.6.4";
 
-  src = fetchFromGitHub {
-    owner = "composer";
-    repo = "composer";
+  # Using `fetchgit` prevent the `composer.lock` file from being stripped out
```
Comment (Ma27) [inline pkgs/development/php-packages/composer/default.nix:9]: s/prevent/prevents/g I think?

Also, not sure if I follow here. What exactly does that mean?

Comment (drupol) [reply pkgs/development/php-packages/composer/default.nix]: I've updated the message, hope it's clearer now.

I also hope we can merge this PR quickly, if possible today.

**Comment by drupol** [comment]
@GrahamcOfBorg build phpPackages.composer

**Comment by drupol** [comment]
Auto merging this myself to avoid people reporting this issue in the upcoming days.

**Comment by yayayayaka** [comment]
It appears the `composer.lock` had been excluded in composer's [`.gitattributes`](https://github.com/composer/composer/commit/0e05c2c68a243799be31485ec17c47b991db7da3#diff-618cd5b83d62060ba3d027e314a21ceaf75d36067ff820db126642944145393eR22). Maybe open a PR in composer to fix that?

**Comment by drupol** [comment]
@seldaek hello !

Would you be ok with that ? If yes, I can provide the PR.

**Comment by Seldaek** [comment]
Sure, I don't really mind either way

**Comment by drupol** [comment]
Thanks! PR is up at https://github.com/composer/composer/pull/11674

**Pull request merged**